carbon-design-system/carbon-addons-iot-react
# fix(SuiteHeader): Adding support for basic browser workflows in Suiteheader actions.

**Summary**

- SuiteHeader navigational actions are now `<a>` elements and not `<button>` elements anymore, so that it is now possible to right click and select "open in a new tab/window" and also ctrl+click to force opening the page in a new tab.
- The rules and effects of the `onRouteChange` custom function are still maintained for left clicks and ctrl+left clicks. Also, the default behavior of existing SuiteHeader navigational actions (whether they open in the same tab or in another tab by default) is also maintained for left clicks (because ctrl+left click forces the navigation to open a new tab). However, for right click -> "open in a new tab/window" action, the `onClick` handlers of navigational buttons are bypassed and the `href` attribute contains the target URL that will open in a new tab/window. The only exception is the "Log Out" button, it is still a button html element that does not allow right click navigational actions. Even in the case of some SuiteHeader menu items that were `<a>` and showed the "open in a new tab", the new tab would open a `about:blank#blocked` because the `href` of these menu items had a `javascript:void(0)` string, and now they have the same URL that the `onClick` handler would use to do the redirect.
- The underlying PAL `Header` used by `SuiteHeader` did not support these browser workflows of `<a>` elements for top level actions that don't have `childContent`. So, a minor change was made in `HeaderAction.jsx` so that it forwards `href`, `rel` and `target` to the `HeaderGlobalAction` carbon component (which basically just renders a `Button`, and carbon `Button` renders an `<a>` if at least `href` is passed to it).

Before:
![image](https://user-images.githubusercontent.com/12237934/155196181-af3f1b76-37ea-4167-8396-10646c0cea05.png)

After:
![image](https://user-images.githubusercontent.com/12237934/155196234-026cf2e6-7a60-4601-bfbd-11c8bdb5b46b.png)


Before:
![image](https://user-images.githubusercontent.com/12237934/155196308-6b886052-3199-47d3-83ad-5b8b495b7db4.png)

After:
![image](https://user-images.githubusercontent.com/12237934/155196370-c85d1775-cabd-4583-af35-63b0d8de1c02.png)

Before:
![image](https://user-images.githubusercontent.com/12237934/155196433-c8c5065c-e497-4c4a-b0d2-251fbbac9501.png)

After:
![image](https://user-images.githubusercontent.com/12237934/155196495-9c7cb1e5-e068-4e3f-8396-fb0dfc19be1e.png)

Log Out button is still a button element:
![image](https://user-images.githubusercontent.com/12237934/155197233-e31f0f3e-0b86-4f18-863e-d70c52c91b35.png)



**Acceptance Test (how to verify the PR)**

- Open the SuiteHeader (and its underlying components) stories and check that you can right click and open the links in a new tab or window.
- Check that you can ctrl+click (cmd+click for Mac) navigational actions and they will open in a new tab.
- Go to the new story ("Navigational actions blocked") and check that clicks and ctrl+clicks don't trigger any navigational actions in the header actions and menu items.
- Check that the Log Out button does not support opening in a new tab from the context menu, nor opening in a new tab through ctrl+click.

**Regression Test (how to make sure this PR doesn't break old functionality)**

- Some tests have been updated to adjust the screen selector code that were relying on `button` elements.
- A new test has been added for `HeaderAction` component to fulfill the minimum coverage requirement.



**Things to look for during review**
